ct/reconciler: Actually enforce the object size limit

wdberkeley opened the pull request:
Oops.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

wdberkeley committed:
ct/reconciler: Actually enforce the object size limit

Oops.

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -456,6 +456,18 @@ reconciler::build_object(
+              max_object_size);
+            break;
+        }
+        ctx.size_budget = max_object_size - current_size;
```
Do we need to worry about overflow here?

wdberkeley replied:
No, because the conditional just above is `!metas.empty() && current_size >= max_object_size`. If it's false, then `metas` is empty, in which case `current_size` must be 0 because no partitions have been successfully added to the object, or `current_size < max_object_size`. If it's true, we break at L467.

andrwng replied:
Is `current_size` actually 0 when `metas` is empty? If so, do we actually need to check `metas.empty()` vs just relying on `current_size >= max_object_size`?

dotnwat replied:
> No, because the conditional just above is !metas.empty() && current_size >= max_object_size. If it's false, then metas is empty, in which case current_size must be 0 because no partitions have been successfully added to the object, or current_size < max_object_size. If it's true, we break at L467.

from the context alone it looks like if `metas` is empty, then there is no particular restriction on the relationship between the values `current_size` and `max_object_size` and it takes quite a bit of context (and continued adherence to a particular behavior) for the relationship that exists now that makes this safe to continue to exist.

i'd favor making the check specific even if its exact current form those checks are redundant.

does that just mean making `&&` be `||` instead?

wdberkeley replied:
Yeah, `||` makes sense.

dotnwat replied:
i was thinking that we'd change the conditional, or after chatting on slack, add a comment of some sort here to explain why we won't get an underflow on the subtraction if metas is empty.

rockwotj reviewed (commented):
Good catch! Can we get a test here for this behavior?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74880-->
<!--e8764cc6-9683-4f6b-a44f-b5f463d8ecb1-->
<!--37ebf964-c0b6-44fa-832a-427af651e57c-->
<!--d5642648-ae3c-44a4-84df-b4ae516e23fd-->
<!--c0f33e7b-0a25-4851-8f96-9c346a4178ed-->
<!--title-->
## Retry command for [Build#74880](https://buildkite.com/redpanda/redpanda/builds/74880)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
```
<!--/body-->

wdberkeley committed:
ct/reconciler: Add multi-source size limit test

I also turned off compression in the single-source test. That test
worked fine, by chance I guess, but compression did affect the
multi-source test.

I added a comment clarifying the size check condition.

wdberkeley commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -456,6 +456,18 @@ reconciler::build_object(
         }
+
+        auto current_size = ctx.builder->file_size();
+        if (!metas.empty() && current_size >= max_object_size) {
```
This is intentional robot, but I will add a comment.

rockwotj commented on src/v/cloud_topics/reconciler/tests/test_utils.h:
```diff
@@ -65,7 +65,7 @@ class fake_source : public source {
             if (
               model::offset_cast(batch.base_offset())
-              < last_reconciled_offset()) {
+              <= last_reconciled_offset()) {
```
I think you need a rebase, this has changed.

rockwotj commented on src/v/cloud_topics/reconciler/tests/reconciler_test.cc:
```diff
@@ -177,6 +179,55 @@ TEST_F(ReconcilerTest, ObjectSizeLimit) {
+      || lro2_round1 < kafka::offset{last_offset_2});
+
+    // Second reconciliation round should process all remaining data.
+    reconcile();
+    EXPECT_EQ(src1->last_reconciled_offset(), kafka::offset{last_offset_1});
+    EXPECT_EQ(src2->last_reconciled_offset(), kafka::offset{last_offset_2});
+    EXPECT_THAT(
+      metastore_next_offset(src1), Optional(kafka::offset{last_offset_1 + 1}));
+    EXPECT_THAT(
+      metastore_next_offset(src2), Optional(kafka::offset{last_offset_2 + 1}));
```
Do we always process src1 first? Is starvation possible here? Do we need to add the `shuffle` back in?

wdberkeley replied:
Oh I missed that we removed the shuffle. I think we want a guaranteed shuffle of the partitions. Otherwise, yes, we can definitely starve partitions. Shuffling helps prevent that.

wdberkeley replied:
After benchmarking, not shuffling is sort of good, in that if we repeatedly process the same partitions first at least we read from the batch cache more often.

wdberkeley commented:
Force push to rebase.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--f60af438-2a48-4a1a-b6a8-769f632c5561-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74945">build#74945</a></summary>


| test_class                           | test_method                     | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                  |
|:-------------------------------------|:--------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests                 | test_rapid_shadow_link_toggling | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74945#019a1745-b502-4dd0-bbe5-921e85029196 | FLAKY         | 15/21    | upstream reliability is '91.65186500888099'. current run reliability is '71.42857142857143'. drift is 20.22329 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling            |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync      | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74945#019a1748-c664-4cb0-a3d1-4355226fbf80 | FLAKY         | 18/21    | upstream reliability is '98.66156787762907'. current run reliability is '85.71428571428571'. drift is 12.94728 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| PartitionMovementTest                | test_empty                      | {"num_to_upgrade": 0}                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/74945#019a1745-b504-4692-a79e-a9dabd9ed1e5 | FLAKY         | 18/21    | upstream reliability is '94.09937888198758'. current run reliability is '85.71428571428571'. drift is 8.38509 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_empty                                |
</details>
<!--/body-->
